pkg/storage/store_snapshot.go

```diff
@@ -53,13 +51,250 @@ const (
+// expected that the implementation that matches the Strategy specified in the
+// snapshot header will always be used.
+type snapshotStrategy interface {
+	// Receive streams SnapshotRequests in from the provided stream and
+	// constructs an IncomingSnapshot.
+	Receive(
+		context.Context, incomingSnapshotStream, SnapshotRequest_Header,
+	) (IncomingSnapshot, error)
+
+	// Send streams SnapshotRequests created from the OutgoingSnapshot in to the
+	// provided stream.
+	Send(
+		context.Context, outgoingSnapshotStream, SnapshotRequest_Header, *OutgoingSnapshot,
+	) error
+
+	// Status provides a status report on the work performed during the
+	// snapshot. Only valid if the strategy succeeded.
+	Status() string
+}
+
+func assertStrategy(header SnapshotRequest_Header, expect SnapshotRequest_Strategy) {
+	if header.Strategy != expect {
+		panic(fmt.Sprintf("expected strategy %s, found strategy %s", expect, header.Strategy))
+	}
+}
+
+// kvBatchSnapshotStrategy is an implementation of snapshotStrategy that streams
+// batches of KV pairs in the BatchRepr format.
+type kvBatchSnapshotStrategy struct {
+	status string
+
+	// Fields used when sending snapshots.
+	batchSize int
+	limiter   *rate.Limiter
+	newBatch  func() engine.Batch
+}
+
+// Send implements the snapshotStrategy interface.
+func (kvSS *kvBatchSnapshotStrategy) Receive(
+	ctx context.Context, stream incomingSnapshotStream, header SnapshotRequest_Header,
+) (IncomingSnapshot, error) {
+	assertStrategy(header, SnapshotRequest_KV_BATCH)
+
+	var batches [][]byte
+	var logEntries [][]byte
+	for {
+		req, err := stream.Recv()
+		if err != nil {
+			return IncomingSnapshot{}, err
+		}
+		if req.Header != nil {
+			err := errors.New("client error: provided a header mid-stream")
+			return IncomingSnapshot{}, sendSnapshotError(stream, err)
+		}
+
+		if req.KVBatch != nil {
+			batches = append(batches, req.KVBatch)
+		}
+		if req.LogEntries != nil {
+			logEntries = append(logEntries, req.LogEntries...)
+		}
+		if req.Final {
+			snapUUID, err := uuid.FromBytes(header.RaftMessageRequest.Message.Snapshot.Data)
+			if err != nil {
+				err = errors.Wrap(err, "invalid snapshot")
+				return IncomingSnapshot{}, sendSnapshotError(stream, err)
+			}
+
+			inSnap := IncomingSnapshot{
+				SnapUUID:   snapUUID,
+				Batches:    batches,
+				LogEntries: logEntries,
+				State:      &header.State,
+				snapType:   snapTypeRaft,
+			}
+			if header.RaftMessageRequest.ToReplica.ReplicaID == 0 {
+				inSnap.snapType = snapTypePreemptive
+			}
+			kvSS.status = fmt.Sprintf("kv batches: %d, log entries: %d", len(batches), len(logEntries))
+			return inSnap, nil
+		}
+	}
+}
+
+// Send implements the snapshotStrategy interface.
+func (kvSS *kvBatchSnapshotStrategy) Send(
+	ctx context.Context,
+	stream outgoingSnapshotStream,
+	header SnapshotRequest_Header,
+	snap *OutgoingSnapshot,
+) error {
+	assertStrategy(header, SnapshotRequest_KV_BATCH)
```

Comment:
why not pass a context in here and use `log.Fatal`?

Reply:
Done.